Sources/NIOHTTP2/HTTP2FrameParser.swift

```diff
@@ -0,0 +1,835 @@
+    /// - If the buffer does not contain an entire frame including payload, it will be cached
+    ///   internally by the decoder, and the next call to `decode(bytes:)` will append any
+    ///   new bytes to that cache in an effort to complete that frame's payload.
+    /// - When enough bytes to decode an entire frame have been accumulated or provided, the
+    ///   resultant frame will be returned. If the entire frame and its payload have not yet
+    ///   been accumulated, `nil` will be returned.
+    /// - If we gather enough bytes to construct a frame, but that frame is invalid or malformed,
+    ///   we consume all the bytes and throw the error specified in RFC 7540.
+    ///
+    /// - note: The HTTP/2 protocol specifies that unknown frame types should be ignored. This
+    ///         implementation will silently collect and drop all bytes associated with frames
+    ///         whose type is unrecognized; it will not throw an error or otherwise provide
+    ///         those bytes for processing elsewhere.
+    ///
+    /// - Parameter bytes: Raw bytes received, ready to decode. On exit, any consumed bytes
+    ///                    will no longer be in the buffer.
+    /// - Returns: A decoded frame type or `nil` if more bytes are needed to complete the frame.
+    /// - Throws: Errors specified by RFC 7540 § 6 for mis-sized frames or incorrect
+    ///           stream/connection association.
+    mutating func decode(bytes: inout ByteBuffer) throws -> HTTP2Frame? {
+        switch self.state {
+        case .idle:
+            return try self.readPotentialFrame(from: &bytes)
+        case .accumulatingData(let header, var currentBytes):
+            let required = header.length - currentBytes.readableBytes
+            if bytes.readableBytes >= required {
+                var slice = bytes.readSlice(length: required)!
+                currentBytes.write(buffer: &slice)
+                return try self.readFrame(withHeader: header, from: &currentBytes)
+            } else {
+                currentBytes.write(buffer: &bytes)  // consume entire input buffer
+                return nil  // no frame (yet)
+            }
+        case .simulatingDataFrames(let header, let padding, let remaining):
+            let streamID = HTTP2StreamID(knownID: Int32(header.rawStreamID & ~0x8000_0000))
+            let flags = HTTP2Frame.FrameFlags(rawValue: header.flags)
+            if bytes.readableBytes < remaining {
+                // ping out a fake DATA frame containing these bytes, without the END_STREAM flag being set
+                let slice = bytes
+                bytes.clear()   // consuming all the bytes
+                self.state = .simulatingDataFrames(originalHeader: header, padding: padding, remainingBytes: remaining - slice.readableBytes)
+                return HTTP2Frame(streamID: streamID, flags: [], payload: .data(.byteBuffer(slice)))
```

Comment:
This branch is not quite right.

It's conceivable that the following statement is true: `bytes.readableBytes < remaining && bytes.readableBytes > (remaining - padding)`. If this occurs, this operation will incorrectly include padding bytes in the DATA frame body, which is not right. I recommend adding a test for this case to verify it misbehaves, and then applying a fix.